File: dotnet/src/PqcStandards/MlKem/Compress.cs
namespace PqcStandards.MlKem;

/// <summary>Compress and Decompress per FIPS 203.</summary>
public static class Compress
{
    /// <summary>Compress: round(2^d / Q * x) mod 2^d.</summary>
    public static int CompressValue(int x, int d)
    {
        // (2^d * x + Q/2) / Q mod 2^d
        long num = ((long)x << d) + Field.Q / 2;
        return (int)(num / Field.Q) & ((1 << d) - 1);
    }

    /// <summary>Decompress: round(Q / 2^d * y).</summary>
    public static int DecompressValue(int y, int d)
    {
        return (int)(((long)Field.Q * y + (1 << (d - 1))) >> d);
    }

    public static int[] CompressPoly(int[] f, int d)
    {
        int[] r = new int[256];
        for (int i = 0; i < 256; i++)
            r[i] = CompressValue(f[i], d);
        return r;
    }

    public static int[] DecompressPoly(int[] f, int d)
    {
        int[] r = new int[256];
        for (int i = 0; i < 256; i++)
            r[i] = DecompressValue(f[i], d);
        return r;
    }
}

File: go/internal/compress/compress.go
// Package compress implements Compress and Decompress per FIPS 203 Equations 4.7 and 4.8.
package compress

import (
	"github.com/liviuepure/PQC-Standards-Implementation/go/internal/field"
)

// Compress computes Compress_d(x) = round((2^d / q) * x) mod 2^d.
// FIPS 203 Eq 4.7.
func Compress(d int, x field.Element) uint16 {
	// round((2^d * x) / q) mod 2^d
	// = ((2^d * x + q/2) / q) mod 2^d  (using integer division for rounding)
	shifted := (uint64(x.Value()) << d) + uint64(field.Q)/2
	result := shifted / uint64(field.Q)
	return uint16(result) & ((1 << d) - 1)
}

// Decompress computes Decompress_d(y) = round((q / 2^d) * y).
// FIPS 203 Eq 4.8.
func Decompress(d int, y uint16) field.Element {
	// round((q * y) / 2^d) = (q * y + 2^(d-1)) / 2^d  (integer division for rounding)
	val := (uint64(field.Q)*uint64(y) + (1 << (d - 1))) >> d
	return field.New(uint16(val))
}
